### PR TITLE
Database dependency injection in planner

### PR DESCRIPTION
**Pull request by grod220**
While working on https://github.com/penumbra-zone/web/pull/1426, it became evident that the state of testing in the wasm code is quite rough given the dependency on the indexeddb global. 

This PR adds a database interface that storage.rs gets passed upon initialization. This allows us to use a mock database that can be used in tests, making our goals around test coverage much more attainable.

### CODE REVIEW

File: packages/wasm/crate/src/database/indexed_db.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            evt.db().create_object_store("TREE_HASHES")?;
+            evt.db().create_object_store("FMD_PARAMETERS")?;
+            evt.db().create_object_store("APP_PARAMETERS")?;
+            evt.db().create_object_store("GAS_PRICES")?;
+        }
+        Ok(())
+    }));
+
+    db_req
+}
```
Comment (grod220) [inline packages/wasm/crate/src/database/indexed_db.rs:67]: Copied over 1:1. We should seek in later PRs to migrate this to MockDb.

File: packages/wasm/crate/src/database/interface.rs
Comment (grod220) [inline packages/wasm/crate/src/database/interface.rs:1]: The core idea of this PR is this interface which allows us to write a mock and write tests that feel much more manageable to write

File: packages/wasm/crate/src/planner.rs [hunk 2]
```diff
@@ -463,7 +462,7 @@ pub async fn plan_transaction(
 
         save_auction_nft_metadata_if_needed(auction_id, &storage, seq).await?;
         let outstanding_reserves: OutstandingReserves =
-            storage.get_auction_oustanding_reserves(auction_id).await?;
```
Comment (grod220) [inline packages/wasm/crate/src/planner.rs:466]: typo

File: packages/wasm/crate/tests/test_mock_db.rs
Comment (grod220) [inline packages/wasm/crate/tests/test_mock_db.rs:1]: An example test suite, but the next PR is to implement a test suite for `get_notes_for_voting()`

**Review by Valentine1898 — APPROVED** [review]
LGTM! 👍

**Comment by TalDerei** [comment]
reviewing the PR now 👍

File: packages/wasm/crate/src/planner.rs [hunk 3]
```diff
@@ -114,7 +115,7 @@ async fn save_unbonding_token_metadata_if_needed(
 /// action renders correctly in the transaction approval dialog.
 async fn save_auction_nft_metadata_if_needed(
     id: AuctionId,
-    storage: &IndexedDBStorage,
+    storage: &Storage<IdbDatabase>,
```
Comment (TalDerei) [inline packages/wasm/crate/src/planner.rs:118]: I really like the generic storage interface!

File: packages/wasm/crate/tests/test_mock_db.rs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+#[wasm_bindgen_test]
+fn test_get_and_put() {
+    let db = MockDb::new();
+    let table_name = "test_table";
+
+    let key = "test_key";
+    let value = AddressIndex::new(1);
+
+    block_on(db.put_with_key(table_name, key, &value)).unwrap();
```
Comment (TalDerei) [inline packages/wasm/crate/tests/test_mock_db.rs:18]: how does `block_on` work?

**Review by TalDerei — APPROVED** [review]
This is a huge improvement to our storage crate, and it's more readable!

File: packages/wasm/crate/src/database/indexed_db.rs [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    let db_req = mock_test_database(db_req).into_future().await;
+
+    let db = db_req.into_future().await?;
+    Ok(db)
+}
+
+// Previous method of testing that requires features in prod code to seed indexed db for tests
+// TODO: Swap out with new MockDb utility for testing
+async fn mock_test_database(mut db_req: OpenDbRequest) -> OpenDbRequest {
+    db_req.set_on_upgrade_needed(Some(|evt: &IdbVersionChangeEvent| -> Result<(), JsValue> {
```
[the remark this answers is not in the document]
Comment (grod220) [reply packages/wasm/crate/src/database/indexed_db.rs:30]: In short, I didn't. By passing in an alternative database, the test suite has full flexibility in instantiating it as it needs.

File: packages/wasm/crate/src/database/mock.rs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+        Ok(result)
+    }
+
+    async fn get_with_index<T, K>(&self, _: &str, _: K, _: &str) -> WasmResult<Option<T>>
+    where
+        T: DeserializeOwned,
+        K: Into<JsValue>,
+    {
+        unimplemented!()
```
[the remark this answers is not in the document]
Comment (grod220) [reply packages/wasm/crate/src/database/mock.rs:80]: Added comments. In short, the person writing the tests that use these methods will have to write them. Figured it's a task for later.

File: packages/wasm/crate/tests/test_mock_db.rs (hunk 4 above)
Comment (grod220) [reply packages/wasm/crate/tests/test_mock_db.rs:18]: I didn't occur to me that I could just define the test fn as async and use await syntax. This was an unnecessary workaround. Will update.

**Pull request merged**